src/main/scala/mesosphere/marathon/core/launcher/InstanceOp.scala

```diff
@@ -86,36 +94,33 @@ object InstanceOp {
         resourceBuilder.build()
       }
 
-      val maybeDestroyVolumes: Option[MesosProtos.Offer.Operation] =
-        if (withDisk.nonEmpty) {
+      val maybeDestroyVolumes: Seq[MesosProtos.Offer.Operation] =
+        withDisk.groupBy(getProviderId).values.map { withDiskResources =>
           val destroyOp =
             MesosProtos.Offer.Operation.Destroy.newBuilder()
-              .addAllVolumes(withDisk.asJava)
-
-          val op =
-            MesosProtos.Offer.Operation.newBuilder()
-              .setType(MesosProtos.Offer.Operation.Type.DESTROY)
-              .setDestroy(destroyOp)
-              .build()
+              .addAllVolumes(withDiskResources.asJava)
 
-          Some(op)
-        } else None
+          MesosProtos.Offer.Operation.newBuilder()
+            .setType(MesosProtos.Offer.Operation.Type.DESTROY)
+            .setDestroy(destroyOp)
+            .build()
+        }.to[Seq]
 
-      val maybeUnreserve: Option[MesosProtos.Offer.Operation] =
+      val maybeUnreserve: Seq[MesosProtos.Offer.Operation] =
         if (withDisk.nonEmpty || reservationsForDisks.nonEmpty) {
-          val unreserveOp = MesosProtos.Offer.Operation.Unreserve.newBuilder()
-            .addAllResources(withoutDisk.asJava)
-            .addAllResources(reservationsForDisks.asJava)
-            .build()
-          val op =
+          (withoutDisk ++ reservationsForDisks).groupBy(getProviderId).values.map { resources =>
```

Comment:
it took me some time to actually understand what is going on here, that we actually need to group resources per provider id because mesos required that...

I actually understood after reading your comment in OfferOperationFactory.scala

Maybe we can change this code to use some named variables or at least add some comments to make it more obvious for the reader?

Maybe we can separate the first step (up to .values) assign it to a variable and add a comment...